File: ruby/lib/apple_purchase_receipt_verifier/roots.rb
# frozen_string_literal: true

require "openssl"
require_relative "roots_data"

module ApplePurchaseReceiptVerifier
  class << self
    # Trust anchors for StoreKit 2 / App Store Server JWS chains.
    #
    # All three published Apple roots, not just the one production chains end
    # at today. Apple documents the third `x5c` element only as "an Apple root
    # certificate" and an App Store Commerce engineer's answer to "always G3?"
    # was "use all Apple Root CAs" — pinning one root would fail closed and
    # silently the day Apple re-anchored a path (PLAN.md D15).
    #
    # @return [Array<OpenSSL::X509::Certificate>] a fresh array of fresh
    #   certificate objects, so a caller mutating it cannot poison another
    #   verifier.
    def apple_jws_roots
      apple_roots
    end

    # Trust anchors for legacy PKCS#7 app-receipt chains. Same three roots, for
    # the same reason.
    #
    # @return [Array<OpenSSL::X509::Certificate>]
    def apple_receipt_roots
      apple_roots
    end

    private

    def apple_roots
      APPLE_ROOT_DER_BASE64.map { |b64| OpenSSL::X509::Certificate.new(b64.unpack1("m0")) }
    end
  end
end

File: python/apple_purchase_receipt_verifier/roots.py
"""Loads the Apple root certificates bundled with this package (copies of
the public roots from https://www.apple.com/certificateauthority/).
Production trust anchors; tests use the shared fixture PKI instead.

Both sets contain all three published Apple roots. Apple deliberately
documents the JWS chain as ending in "an Apple root certificate" (not a
specific one) and its guidance is to trust every root on the PKI page, so
anchoring on a single root would break silently if Apple re-anchored a
path — see PLAN.md D15."""

from pathlib import Path

from cryptography import x509

_CERTS = Path(__file__).parent / "certs"


def _load(name: str) -> x509.Certificate:
    return x509.load_der_x509_certificate((_CERTS / name).read_bytes())


def _all_roots() -> "list[x509.Certificate]":
    return [
        _load("AppleIncRootCertificate.cer"),
        _load("AppleRootCA-G2.cer"),
        _load("AppleRootCA-G3.cer"),
    ]


def apple_jws_roots() -> "list[x509.Certificate]":
    """Trust anchors for StoreKit 2 / App Store Server JWS chains.
    Production chains currently end at Apple Root CA - G3."""
    return _all_roots()


def apple_receipt_roots() -> "list[x509.Certificate]":
    """Trust anchors for legacy PKCS#7 app-receipt chains.
    Production chains currently end at the Apple Inc. Root CA."""
    return _all_roots()
